src/core/geometry/qgscompoundcurve.cpp

```diff
@@ -906,6 +907,101 @@ QVector< QPair<int, QgsVertexId> > QgsCompoundCurve::curveVertexId( QgsVertexId
   return curveIds;
 }
 
+bool QgsCompoundCurve::convertVertex( QgsVertexId position )
+{
+
+  // First we find out the sub-curves that are contain that vertex.
+
+  // If there is more than one, it means the vertex was at the beginning or end
+  // of an arc, which we don't support.
+
+  // If there is exactly one, we may either be on a LineString, or on a CircularString.
+
+  // If on CircularString, we need to check if the vertex is a CurveVertex (odd index).
+  // If so, we split the subcurve at vertex -1 and +1, , drop the middle part and insert a LineString/CircularString
+  // instead with the same points.
+  
+  // At the end, we call condenseCurves() to merge successible line/circular strings
+
+  QVector< QPair<int, QgsVertexId> > curveIds = curveVertexId( position );
+
+  // We cannot convert points at start/end of subcurves
+  if ( curveIds.length() != 1 )
+    return false;
+
+  int curveId = curveIds[0].first;
+  QgsVertexId subVertexId = curveIds[0].second;
+  QgsCurve *curve = mCurves[curveId];
+
+  // We cannot convert first/last point of curve
+  if ( subVertexId.vertex == 0 || subVertexId.vertex == curve->numPoints() - 1 )
+    return false;
+
+  // TODO factorize circularString and lineString as logic is almost the same
+  if ( const QgsCircularString *circularString = dynamic_cast<const QgsCircularString *>( curve ) )
```

Comment:
qgsgeometry_cast (micro optimization and RTTI independency which is absolutely irrelevant, because we heavily depend on RTTI everywhere -- so let's just call it "best practice" :stuck_out_tongue_winking_eye: )